MapCell

alevy opened the pull request:
"Fixes" #178 by introducing a new type, `MapCell` that does not need to move the value out on a `map`.

As noted in the issue, this improves performance significantly at the cost of a word of storage in case the data value is a reference (because `TakeCell` is able to take advantage of the implicit `NonZero` optimization in `Option`).

Questions:

1. What should these be called?

2. Should `TakeCell` and `MapCell` implement a common trait (they currently have identical interfaces)

3. Should `TakeCell` _only_ admit references?

bradjc commented:
1. Well we can't use RefCell. Pat suggested `StaticRefCell`. And then `ObjectCell`?
2. Maybe? I guess it doesn't hurt.
3. Yes.

Also there are a couple "TakeCell"s in your comments.

brghena commented:
My suggestions:

1. I propose `RefShare` and `ValShare` (possibly `ReferenceShare` and `ValueShare`)
2. Is there any circumstance where you would accept either of them and not care which? It seems like there will not be if the answer to **3** is yes.
3. Yes

alevy commented:
1. I like `RefShare` and `ValShare`. I think @phil-levis had wanted to make sure the names are descriptive, so maybe he has thoughts.

2. @brghena the only circumstance so far is the test I wrote, which just duplicates code for now... There is maybe some advantage from a documentation perspective and finally I could imagine library code that might be agnostic. But I think the current answer is no real use-case.

3. Seems like the consensus is yes, so that's an easy one.

@bradjc will fix comments

bradjc commented:
My only issue with using `Share` (or anything not Cell) is that if the first line of the comment is "this is like a Rust `Cell`, but..." then existing rust developers have to learn this new term we made up. If the types are different enough that linking them to Cell is not helpful, then great, but if not, having the lineage of being similar to `Cell`, `RefCell`, `UnsafeCell` (all in standard Rust) should make it easier for people (particularly rust developers) to get started with tock.

phil-levis commented:
I like Ref and Val. I'm a bit leery of Share because it indicates why we use it, not what it does. But I could live it it. How about RefPrison and ValPrison? ;)

alevy commented:
This is complete once the tests pass. I did not rename the types yet, since there isn't consensus on the names. However, TakeCell now only admits mutable references. I replaced most occurrences of TakeCell with either `Cell<Option<T>>` or `MapCell`.

Worth testing a few applications before merging, since it's plausible I introduced some bugs in the translation (it was a lot of refactoring).

ppannuto commented on capsules/src/console.rs:
```diff
@@ -204,15 +205,18 @@ impl<'a, U: UART> Client for Console<'a, U> {
         self.tx_buffer.replace(buffer);
-        self.in_progress.take().map(|appid| {
+        self.in_progress.get().map(|appid| {
+            self.in_progress.set(None);
```
@brghena and I were talking about this. It seems like this is a downgrade from a usage perspective. We've gone from `.map(||)` simply doing the right thing to `.get().map(||)` + don't forget to `set(None)`.

Could `in_progress` be in a MapCell? Or do we need yet a third Cell type?

alevy replied:
It could be a `MapCell`. This is i think the software engineering trade-off we talked about offline. Happy to change this to a MapCell as an example

brghena replied:
What problem would come up if the user forgets to set(None)? It seems like if that is going to be a problem, paying one word to avoid it everywhere is worth it.

brghena replied:
On second look, it seems like needing to .set(None) is not the normal case.

I'd still very much prefer a common paradigm where we can just call .map() on anything rather than having some .get().map() and some .map()

ppannuto replied:
Agreed on both fronts, but neither are sufficient to stand in the way of merging this as-is right now IMO. I think we should pull the trigger.

ppannuto replied:
[well, rustfmt should be fixed so travis goes green, but yeah]

ppannuto commented:
Otherwise happy to report that things seem to be largely working. Tested a mixture of blink, prinf(long), sensors, accel-leds, rot13_(client/server) without issue on hail.
